### PR TITLE
join filter closure and constant join lookups

### PR DESCRIPTION
**Pull request by max-hoffman**
This PR adds a set of join planning improvements.

1. Table aliases can accept multi column indexes

We have never been able to choose a multi-expression range scan through table aliases.

Before:
```sql
tmp2> explain select * from t alias where a = 1 and b = 1 and c = 1;
+-----------------------------------------------------------+
| plan                                                      |
+-----------------------------------------------------------+
| Filter                                                    |
|  ├─ (((alias.a = 1) AND (alias.b = 1)) AND (alias.c = 1)) |
|  └─ TableAlias(alias)                                     |
|      └─ IndexedTableAccess(t)                             |
|          ├─ index: [t.a]                                  |
|          ├─ filters: [{[1, 1]}]                           |
|          └─ columns: [a b c]                              |
+-----------------------------------------------------------+
```

After:
```sql
tmp2> explain select * from t alias where a = 1 and b = 1 and c = 1;
+-----------------------------------------------------------+
| plan                                                      |
+-----------------------------------------------------------+
| Filter                                                    |
|  ├─ (((alias.a = 1) AND (alias.b = 1)) AND (alias.c = 1)) |
|  └─ TableAlias(alias)                                     |
|      └─ IndexedTableAccess(t)                             |
|          ├─ index: [t.a,t.b,t.c]                          |
|          ├─ filters: [{[1, 1], [1, 1], [1, 1]}]           |
|          └─ columns: [a b c]                              |
+-----------------------------------------------------------+
```

This has silently been impacting join performance in particular, where table aliases are more common. This is a small change but I'd expect this to have a broad positive impact for customers.

2. Join equivalence closure

A join like `select * from xy join uv on x = u join ab on u = a` has two initial join edges, `x = u` and `u = a`. Those edges create expression groupings `xy x uv`, `uv x ab`, `xy x uv x ab`. There misses a transitive edge, `x = a`, with a corresponding join group `xy x ab`. We should generate plans for most transitive edges now (transitive edges in apply joins are harder).

For joins with many tables this will unlock many potential join paths.

3. Use functional dependencies to find more lookup and merge joins

We can use constants and aggregated equivalency sets (equal filters) to be more aggressive with lookup join selection. Previously we only searched the current join ON equal conditions for expressions that match an index prefix for a lookup join, but constants are also valid lookup keys.

Refer to https://github.com/dolthub/dolt/issues/5993 and https://github.com/dolthub/dolt/issues/3797 for in-depth examples.

4. Use functional dependencies to do better lookup join costing

Even though we do not have index statistics, we can still use functional dependencies on indexes to detect whether a lookup will have MAX_1_ROW. Two examples where we can detect MAX_1_ROW: our lookup index is the primary key, and our lookup key provides a constant or equals expression for every pk column; our lookup index is unique, our lookup key has constants or equal expressions for every column, and we can prove that every key expression is non-nullable.

MAX_1_ROW lookups are a rare binary condition, most of the time selectivity is in the continuous range 0-1. When they do occur they are usually the most efficient access pattern. Many of the test changes from HASH_JOIN or MERGE_JOIN to LOOKUP_JOIN are a result of this improvement. The issues linked above in (3) have practical examples.

### CODE REVIEW

File: sql/memo/coster.go [hunk 1]
```diff
@@ -205,20 +205,49 @@ func (c *coster) costDistinct(_ *sql.Context, n *Distinct, _ sql.StatsReader) (f
...
+				notNull = notNull.Union(onCols)
+			}
+		}
+		// from the perspective of the secondary table, lookup keys
+		// will be constant
+		constCols = constCols.Union(onCols)
+	}
+
+	fds := sql.NewLookupFDs(l.Parent.Right.RelProps.FuncDeps(), l.Index.ColSet(), notNull, constCols, joinFds.Equiv())
+	max1Row := fds.HasMax1Row()
```
Comment (zachmu) [inline sql/memo/coster.go:247]: simplify to `if fds.HasMax1Row() { return 1 }`

File: sql/memo/join_order_builder.go [hunk 2]
```diff
@@ -182,6 +180,103 @@ func (j *joinOrderBuilder) populateSubgraph(n sql.Node) (vertexSet, edgeSet, *Ex
...
 }
 
+// ensureClosure adds the closure of all transitive equivalency groups
+// to the join tree. Each transitive edge will add an inner edge, filter,
+// and join group that inherit join type and tree depth from the original
+// join tree.
+func (j *joinOrderBuilder) ensureClosure(grp *ExprGroup) {
+	fds := grp.RelProps.FuncDeps()
+	for _, set := range fds.Equiv().Sets() {
+		for col1, ok1 := set.Next(1); ok1; col1, ok1 = set.Next(col1 + 1) {
```
Comment (zachmu) [inline sql/memo/join_order_builder.go:190]: Not really part of this PR, but `Next()` is usually part of an iterator pattern that doesn't usually take an arg

You probably want this to be called `Get`

Also use `hasNext` instead of `ok1` here

File: sql/memo/join_order_builder.go [hunk 3]
```diff
@@ -182,6 +180,103 @@ func (j *joinOrderBuilder) populateSubgraph(n sql.Node) (vertexSet, edgeSet, *Ex
...
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// makeTransitiveEdge constructs a new join tree edge and memo group
+// on an equality filter between two columns.
+func (j *joinOrderBuilder) makeTransitiveEdge(col1, col2 sql.ColumnId) {
```
Comment (zachmu) [inline sql/memo/join_order_builder.go:229]: Would be great to see unit tests of this function and ensureClosure

Not to test it per se, but to tighten up and make explicit the contract. Basically the best kind of documentation you can write.

File: sql/memo/rel_props.go [hunk 4]
```diff
@@ -52,13 +53,257 @@ func newRelProps(rel RelExpr) *relProps {
...
+func denormIdxExprs(table string, idx sql.Index) []string {
+	denormExpr := make([]string, len(idx.Expressions()))
+	for i, e := range idx.Expressions() {
+		parts := strings.Split(e, ".")
+		denormExpr[i] = strings.ToLower(fmt.Sprintf("%s.%s", table, parts[1]))
+	}
+	return denormExpr
+}
+
+func (p *relProps) populateFds() {
```
Comment (zachmu) [inline sql/memo/rel_props.go:78]: Same comment about unit tests for this function

Doesn't need to be extensive, but having ~1 unit test for each case in the switch statement is going to make this a lot easier to understand for the next guy

File: sql/memo/select_hints_test.go [hunk 5]
```diff
@@ -172,9 +171,14 @@ func TestOrderHintBuilding(t *testing.T) {
 				10: testVertexSet(0, 1, 2),    // ab x xy x pq
 				11: testVertexSet(3),          // uv
 				14: testVertexSet(0, 1, 2, 3), // ab x xy x pq x uv
-				15: testVertexSet(1, 2),       // xy x pq
-				16: testVertexSet(2, 3),       // pq x uv
-				17: testVertexSet(1, 2, 3),    // xy x pq x uv
+				18: testVertexSet(0, 2),       // ab x pq
```
Comment (zachmu) [inline sql/memo/select_hints_test.go:174]: On the same theme, these would be more useful if the tests were defined in terms of the comments and you did translation via some function

As these stand, they're brittle: they're going to break when we fix a bug or make other changes, and it's going to be very hard for someone to figure out why

For example it's super non-obvious why the group ids in these expected values aren't contiguous. If the answer is "that's a detail of the way we build the data structure" fine, but that should be abstracted away from how we build the test expectation data in that case. This kind of definition leads to people just regenerating the tests because they don't understand how to construct them.

**Review by zachmu — APPROVED** [review]
These are all good changes and I don't have any real objections, just a couple general comments:

* This PR has a lot of (I think) separable changes, and it would be easier to review well if they were each their own PR
* Probably past time to invest more in the tests here now that the interfaces are settling. See specific PR comments.

File: sql/analyzer/indexed_joins.go [hunk 6]
```diff
@@ -261,11 +256,9 @@ func addLookupJoins(m *memo.Memo) error {
 			for _, on := range conds {
 				filters := memo.SplitConjunction(on)
```
[the remark this answers is not in the document]
Comment (max-hoffman) [reply sql/analyzer/indexed_joins.go:257]: on isn't a `sql.Expression`, its a `memo.ScalarExpr`

File: sql/memo/select_hints_test.go (hunk 5 above)
Comment (max-hoffman) [reply sql/memo/select_hints_test.go:174]: yeah i'll rewrite, i'm going to end up reversing this and making these contiguous again soon i think
